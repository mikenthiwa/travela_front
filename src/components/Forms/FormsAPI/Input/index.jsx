import React, { PureComponent, Fragment } from 'react';
import { PropTypes } from 'prop-types';
import {
  HtmlInput,
  DropdownSelect,
  DateInput,
  ButtonToggler,
  NumberInput,
  filterDropdownSelect,
  CheckBox,
  TextArea,
  TagsInput,
  MultipleChoiceDropdown,
  SelectChoiceDropdown,
  OnboardingButtonToggler,
} from './InputFields';
import createEventHandlersFor from '../formEventHandlers';
import './_input.scss';
import './InputFields/Date.scss';
import { RealFormContext } from '../FormContext/FormContext';

class Input extends PureComponent {

  newProps = {};

  getInputType() {
    let { name, type } = this.props;
    const { targetForm, validatorName } = this;
    // get event handlers for the form in this context, for its input named 'name'
    const eventHandlers = createEventHandlersFor(targetForm, name, validatorName);
    return this.switchTypeWithProps(type, eventHandlers);
  }

    switchTypeWithProps = (type, eventHandlers) => {
      const { onChange, minDate, maxDate } = this.props;
      switch (type) {
      case 'button-toggler':
        this.newProps = {
          ...this.props,
          onChange: onChange ||  eventHandlers.handleSelectTogglerOpt
        };
        return ButtonToggler;
      case 'onboarding-button-toggler':
        this.newProps = {
          ...this.props,
          onChange: onChange ||  eventHandlers.handleSelectTogglerOpt
        };
        return OnboardingButtonToggler;
      case 'date':
        this.newProps  = {
          ...this.props,
          minimumDate: minDate,
          maximumDate: maxDate,
          onChange: onChange || eventHandlers.handleSelectDate,
          onBlur: eventHandlers.handleInputBlur
        };
        return DateInput;
      case 'dropdown-select':
      case 'filter-dropdown-select':
        this.newProps= {
          ...this.props,
          onChange: onChange || eventHandlers.handleSelectDropdown,
        };
        return this.switchDropdownInputTypes(type);
      case 'checkbox':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleCheckBoxChange
        };
        return CheckBox;
      case 'tags':
        this.newProps = {
          ...this.props,
          handleDelete: onChange || eventHandlers.handleTagDeleted,
          handleAddition: onChange || eventHandlers.handleTagAdded,
          handleInputBlur: eventHandlers.handleTagAdded
        };
        return TagsInput;
      case 'multiple-choice-dropdown':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleInputChange,
          onBlur: eventHandlers.handleInputBlur
        };
        return MultipleChoiceDropdown;
      case 'select-choice-dropdown':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleInputChange,
          onBlur: eventHandlers.handleInputBlur
        };
        return SelectChoiceDropdown;
      default:
        this.newProps = {
          ...this.props,
          onBlur: eventHandlers.handleInputBlur,
          onChange: onChange ||  eventHandlers.handleInputChange
        };
        return this.switchTextBasedInputTypes(type);
      }
    }

    switchTextBasedInputTypes(type) {
      switch (type) {
      case 'text':
      case 'password':
      case 'email':
        return HtmlInput;
      case 'number':
        return NumberInput;
      case 'textarea':
        return TextArea;
      }
    }

    switchDropdownInputTypes(type) {
      switch (type) {
      case 'filter-dropdown-select':
        return filterDropdownSelect;
      default:
        return DropdownSelect;
      }
    }

    labelNote(note) { // a small message tied to an inputs label
      return note ? (
        <span>
        &nbsp;
          { note }
        </span>
      ) : null;
    }


    renderAsterisk(){
      /* This API assumes that all inputs are required and therefore a red asterisk is displayed
         *  but this is not enforced with html
         * To remedy the former, if one has a field that should be optional,
         * they have to explicitly pass 'required: false' into customProps
         * This method checks for this and will display the asterisk appropriately.
         */

      const { required, labelNote } = this.props;
      let requiredInput = required === undefined ? true : required;
      return (
        <Fragment>
          {requiredInput && ( <span style={{color: 'red'}}> * </span> ) }
          {this.labelNote(labelNote)}
        </Fragment>
      );
    }

    render() {
      return (
        <RealFormContext.Consumer>
          {({ errors, values, targetForm, validatorName }) => {
            this.targetForm = targetForm;
            this.validatorName = validatorName;

            let { name, label, className } = this.props;
            const value = values ? values[name]: '';
            const error = errors ? errors[name]: '';
            let customClass = className ? className : '';
            // switch input types into InputElement
            const InputElement = this.getInputType();
            return (
              <div
                className={`form-input ${customClass} ${error ? 'error' : ''}`}>
                <label htmlFor={name}>
                  {label}
                  {this.renderAsterisk()}
                </label>
                <InputElement value={value} error={error} {...this.newProps} />
                <span className="error">
                  {error}
                </span>
              </div>
            );
          }}
        </RealFormContext.Consumer>
      );
    }
}

Input.contextTypes = {
  errors: PropTypes.object,
  values: PropTypes.object,
  targetForm: PropTypes.object,
  validatorName: PropTypes.string
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.oneOfType([
    PropTypes.func, PropTypes.object
  ]),
  labelNote: PropTypes.string,
  className: PropTypes.string,
};

Input.defaultProps = {
  labelNote: '',
  className: '',
  onChange: null
};

export default Input;
