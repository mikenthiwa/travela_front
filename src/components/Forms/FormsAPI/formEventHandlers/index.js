// This is used by the  Input component to generate event handlers
/*
  eventHandlers will update the form and validate the input based on the validate logic in the form.
*/
import * as handlerCreators from './eventHandlerCreators';
import getDefaultBlanksValidatorFor from '../formValidator';


const createEventHandlersFor = (targetForm, targetField, validatorName) => {
  // creates every event handler for a form
  let eventHandlers = {};
  const _handlerCreators = {
    handleSelectDropdown: handlerCreators.getDropdownHandler,
    handleSelectTogglerOpt: handlerCreators.getTogglerHandler,
    handleInputBlur: handlerCreators.getOnBlurHandler,
    handleSelectDate: handlerCreators.getDateHandler,
    handleInputChange: handlerCreators.getInputChangeHandler,
    handleCheckBoxChange: handlerCreators.getCheckBoxHandler,
    handleTagAdded: handlerCreators.getTagsAddHandler,
    handleTagDeleted: handlerCreators.getTagsDeleteHandler
  };
  // call the creators in _handlerCreators to create the handlers
  // for the target form and store them in eventHandlers
  const validator = targetForm[validatorName || 'validate'] ||
      getDefaultBlanksValidatorFor(targetForm);
  for (let handlerName of Object.keys(_handlerCreators)) {
    eventHandlers[handlerName] = _handlerCreators[handlerName](
      targetForm,
      validator,
      targetField
    );
  }
  return eventHandlers;
};

export default createEventHandlersFor;
