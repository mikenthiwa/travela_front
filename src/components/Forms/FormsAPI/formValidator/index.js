
function validate(field)  {
  const targetForm = this;
  let { values, errors, optionalFields, trips } = targetForm.state;
  [errors, values] = [{ ...errors }, { ...values }];
  let hasBlankFields = false;
  let sameOriginDestination = true;

  hasBlankFields = Object.keys(values)
    .some(key => !values[key] && !isOptional(key, optionalFields));

  let strArray;
  let index;
  if (field != undefined){
    const str = field;
    strArray = str.split('-'); 
    index  = strArray[1];
  }
  
  if((values[`origin-${index}`] === values[`destination-${index}`])){
    if((values[`origin-${index}`] != '' || values[`destination-${index}`] != '')){
      errors[`origin-${index}`] = 'same origin and destination';
      errors[`destination-${index}`] = 'same origin and destination';
      targetForm.setState(prevState => ({
        ...prevState, errors, hasBlankFields, sameOriginDestination }));
    }
  }else {
    sameOriginDestination = false;
    errors[`origin-${index}`] = '';
    errors[`destination-${index}`] = '';
    targetForm.setState(prevState => ({
      ...prevState, errors, hasBlankFields, sameOriginDestination}));
  }
  // when removing a trip on edit 
  if(!index && targetForm.props.requestOnEdit && 
    (targetForm.props.requestOnEdit.trips.length !== trips.length)){
    targetForm.setState(prevState =>({
      ...prevState, sameOriginDestination: false
    }));
  }

  if (!field){
    targetForm.setState({hasBlankFields});
    return !hasBlankFields;
  }

  errors[field] = !values[field] && !isOptional(field, optionalFields)
    && 'This field is required';

  targetForm.setState(prevState => ({ ...prevState, errors, hasBlankFields}));
  return !hasBlankFields;
}

export const isOptional = (field, optionalFields) => {
  return optionalFields && optionalFields
    .some(expr => new RegExp(expr).test(field));
};


const getDefaultBlanksValidatorFor = (targetForm) => {
  return validate.bind(targetForm);
};

export default getDefaultBlanksValidatorFor;
