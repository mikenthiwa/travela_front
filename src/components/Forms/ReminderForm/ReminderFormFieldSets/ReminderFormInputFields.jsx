import React, { Component, Fragment } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import InputRenderer from '../../FormsAPI';
import addButton from '../../../../images/add.svg';
import deleteBtnRed from '../../../../images/delete.svg';
import AddInputRowButton from '../../AddInputRowButton';
import * as formMetaData from '../../FormsMetadata/ReminderForm/index';

class ReminderFormInputFields extends Component {
  state = {};


  processFormMetadata = (reminderIndex) => {
    const { dropdownSelectOptions, inputLabels } = formMetaData;

    const { props : { values} } = this;
    const value = values[`date-${reminderIndex}`];

    const period =dropdownSelectOptions.period.map((period) => (
      parseInt(value) === 1 || !value ? period : `${period}s`
    ));

    return {dropdownSelectOptions: {...dropdownSelectOptions, period}, inputLabels};
  };

  selectEnabledTemplates = () => {
    const { templates } = this.props;
    return templates
      .map(template => (
        { value: `${template.id}`,
          label: template.name
        }
      ));
  };

  renderTemplateDropDown(reminderIndex) {
    const { renderInput } = this.inputRenderer;
    const {
      templates,
      currentPage,
      loading,
      pageCount,
      fetchAllEmailTemplates,
      onReminderTemplateChange } = this.props;
    const choices = this.selectEnabledTemplates();

    return renderInput(`reminderTemplate-${reminderIndex}`, 'dropdown-select', {
      label: `Reminder ${reminderIndex + 1}`,
      choices,
      templatesCount: choices.length,
      currentPage: currentPage,
      loading: loading,
      pageCount: pageCount,
      fetchAllEmailTemplates: fetchAllEmailTemplates,
      onChange: value => onReminderTemplateChange(value, reminderIndex),
    });
  }

  renderDatePeriodInputGroup(reminderIndex) {
    const { renderInput } = new InputRenderer(this.processFormMetadata(reminderIndex));
    const { onReminderPeriodChange, onInputChange } = this.props;
    return (
      <Fragment>
        <div className="dropdown-input-group">
          {renderInput(`date-${reminderIndex}`, 'number', { min: 1, id: reminderIndex, onChange: onInputChange })}
          {renderInput(`period-${reminderIndex}`, 'dropdown-select', {
            className: 'period',
            onChange: value => onReminderPeriodChange(value, reminderIndex),
          })}
          <p className="expiry-text-mobile">To Expiry</p>
        </div>
        <p className="expiry-text-desktop">To Expiry</p>
      </Fragment>
    );
  }

  renderRemoveReminderButton(reminderIndex) {
    const { removeReminder } = this.props;
    return (
      <button
        type="button"
        className="delete-icon"
        onClick={() => removeReminder(reminderIndex)}
      >
        <img src={deleteBtnRed} alt="clicked" className="addsvg" />
      </button>
    );
  }

  renderReminder(reminderIndex) {
    return (
      <div key={reminderIndex} className="reminders-fieldset">
        {this.renderTemplateDropDown(reminderIndex)}
        {this.renderDatePeriodInputGroup(reminderIndex)}
        {reminderIndex > 0 && this.renderRemoveReminderButton(reminderIndex)}
      </div>
    );
  }

  renderConditionInput() {
    const { renderInput } = this.inputRenderer;
    return renderInput('conditionName', 'text');
  }

  renderReminderInputSets(totalReminders) {
    const reminders = [];
    for (let i = 0; i < totalReminders; i++) {
      reminders.push(this.renderReminder(i));
    }
    return reminders;
  }

  renderAddReminderButton() {
    const { addNewReminder } = this.props;
    return (
      <AddInputRowButton
        text="Add a Reminder"
        addRowHandler={addNewReminder}
        wrapperClassName="add-reminder__border-dashed"
        buttonClassName="add-reminder__button"
      />
    );
  }

  render() {
    this.inputRenderer = new InputRenderer(formMetaData);
    const { totalReminders } = this.props;
    return (
      <fieldset>
        {this.renderConditionInput()}
        {this.renderReminderInputSets(totalReminders)}
        {this.renderAddReminderButton()}
      </fieldset>
    );
  }
}

ReminderFormInputFields.propTypes = {
  totalReminders: PropTypes.number.isRequired,
  addNewReminder: PropTypes.func.isRequired,
  removeReminder: PropTypes.func.isRequired,
  onInputChange: PropTypes.func.isRequired,
  onReminderTemplateChange: PropTypes.func.isRequired,
  onReminderPeriodChange: PropTypes.func.isRequired,
  templates: PropTypes.array,
  currentPage: PropTypes.number,
  loading: PropTypes.bool.isRequired,
  values: PropTypes.object.isRequired,
  pageCount: PropTypes.number,
  fetchAllEmailTemplates: PropTypes.func.isRequired,
};

ReminderFormInputFields.defaultProps = {
  templates: [],
  currentPage: 1,
  pageCount: 100,
};

export default ReminderFormInputFields;
