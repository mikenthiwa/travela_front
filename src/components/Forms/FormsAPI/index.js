import React, {Component} from 'react';
import Input from './Input';
import FormContext from './FormContext/FormContext';
import getDefaultBlanksValidatorFor from './formValidator';

class InputRenderer {
  constructor(formMetadata) {
    this.formMetadata = formMetadata;
  }

  switchProps(name, type, inputProps) {
    const formMetadata = this.formMetadata || {};
    switch (type) {
    case 'filter-dropdown-select':
      inputProps.choices = this.formMetadata.dropdownSelectOptions[name];
      return inputProps;
    case 'dropdown-select':
      inputProps.choices = (formMetadata.dropdownSelectOptions || {})[name.split('-')[0]];
      return inputProps;
    case 'button-toggler':
      inputProps.choices = (formMetadata.buttonToggleOptions || {})[name];
      return inputProps;
    case 'multiple-choice-dropdown':
      return inputProps;
    default:
      return inputProps;
    }
  }

  renderInput = (name, type, customProps) => {
    const selection = customProps && customProps.selection;
    const labelKey = name.split('-')[0];
    const fieldName = (labelKey.match('arrivalDate') && (selection && !selection.match('return')))
      ? 'leavingDate' : labelKey;
    let inputProps = {
      name,
      type,
      label: this.formMetadata.inputLabels[fieldName].label,
      labelNote: customProps ? customProps.labelNote :
        this.formMetadata.inputLabels[fieldName].note,
      autoComplete: 'off'
    };
    customProps ? inputProps['data-parentid'] = customProps.parentid : null;
    inputProps = this.switchProps(name, type, inputProps);
    inputProps = {
      ...inputProps,
      ...customProps
    };
    return createInput(inputProps);
  }
}

const createInput = props => <Input {...props} />;

export {Input, FormContext, getDefaultBlanksValidatorFor};
export default InputRenderer;
