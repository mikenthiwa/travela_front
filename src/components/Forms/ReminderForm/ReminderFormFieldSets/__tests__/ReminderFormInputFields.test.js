import React from 'react';
import ReminderFormInputFields from '../ReminderFormInputFields';
import { Input } from '../../../FormsAPI';

const props = {
  totalReminders: 3,
  addNewReminder: jest.fn(),
  removeReminder: jest.fn(),
  onInputChange: jest.fn(),
  onReminderTemplateChange: jest.fn(),
  onReminderPeriodChange: jest.fn(),
  templates: [{
    id: '1',
    name: 'Email Reminder'
  }],
  currentPage: 3,
  loading: false,
  pageCount: 4,
  values: {
    'date-0': 1,
    'date-1': 2,
    'date-2': 10
  },
  fetchAllEmailTemplates: jest.fn(),
};
describe('ReminderFormInputFields', () => {
  const setup = (props) => shallow(<ReminderFormInputFields {...props} />);
  it('should render correctly', () => {
    const wrapper = setup(props);
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the condition Input', () => {
    const wrapper = setup(props);
    const conditionInput = wrapper.find(Input).at(0);
    expect(conditionInput.length).toBe(1);
  });

  it('should render the correct number of reminder inputs', () => {
    const wrapper = setup(props);
    const remindersFieldset = wrapper.find('.reminders-fieldset');
    const totalInputs = remindersFieldset.find(Input);
    expect(totalInputs.length).toBe(9); // 3 input fields per reminder
  });

  it('should render the correct singular or plural form of dates', () => {
    const wrapper = setup(props);
    const singular = ['Day', 'Week', 'Month', 'Year'];
    const plural = singular.map((period) => `${period}s`);
    expect(wrapper.find('Input[name="period-0"]').props().choices).toEqual(singular);
    expect(wrapper.find('Input[name="period-1"]').props().choices).toEqual(plural);
    expect(wrapper.find('Input[name="period-2"]').props().choices).toEqual(plural);
  });

  it('should simulate onchange and onclick when they are called', () => {
    const wrapper = setup(props);
    const event = { preventDefault: jest.fn()};
    wrapper.find('[name="reminderTemplate-2"]').simulate('change', event);
    expect(props.onReminderTemplateChange).toBeCalled();
    wrapper.find('[name="period-0"]').simulate('change', event);
    expect(props.onReminderPeriodChange).toBeCalled();
    wrapper.find('.delete-icon').at(0).simulate('click');
    expect(props.removeReminder).toBeCalled();
  });
});
