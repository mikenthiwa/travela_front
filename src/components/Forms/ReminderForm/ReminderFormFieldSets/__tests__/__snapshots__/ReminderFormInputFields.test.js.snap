// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderFormInputFields should render correctly 1`] = `
<fieldset>
  <Input
    autoComplete="off"
    className=""
    label="Condition Name"
    labelNote=""
    name="conditionName"
    onChange={null}
    type="text"
  />
  <div
    className="reminders-fieldset"
    key="0"
  >
    <Input
      autoComplete="off"
      choices={
        Array [
          Object {
            "label": "Email Reminder",
            "value": "1",
          },
        ]
      }
      className=""
      currentPage={3}
      fetchAllEmailTemplates={[Function]}
      label="Reminder 1"
      labelNote=""
      loading={false}
      name="reminderTemplate-0"
      onChange={[Function]}
      pageCount={4}
      templatesCount={1}
      type="dropdown-select"
    />
    <div
      className="dropdown-input-group"
    >
      <Input
        autoComplete="off"
        className=""
        id={0}
        label="Date"
        labelNote=""
        min={1}
        name="date-0"
        onChange={[Function]}
        type="number"
      />
      <Input
        autoComplete="off"
        choices={
          Array [
            "Day",
            "Week",
            "Month",
            "Year",
          ]
        }
        className="period"
        label=""
        labelNote=""
        name="period-0"
        onChange={[Function]}
        type="dropdown-select"
      />
      <p
        className="expiry-text-mobile"
      >
        To Expiry
      </p>
    </div>
    <p
      className="expiry-text-desktop"
    >
      To Expiry
    </p>
  </div>
  <div
    className="reminders-fieldset"
    key="1"
  >
    <Input
      autoComplete="off"
      choices={
        Array [
          Object {
            "label": "Email Reminder",
            "value": "1",
          },
        ]
      }
      className=""
      currentPage={3}
      fetchAllEmailTemplates={[Function]}
      label="Reminder 2"
      labelNote=""
      loading={false}
      name="reminderTemplate-1"
      onChange={[Function]}
      pageCount={4}
      templatesCount={1}
      type="dropdown-select"
    />
    <div
      className="dropdown-input-group"
    >
      <Input
        autoComplete="off"
        className=""
        id={1}
        label="Date"
        labelNote=""
        min={1}
        name="date-1"
        onChange={[Function]}
        type="number"
      />
      <Input
        autoComplete="off"
        choices={
          Array [
            "Days",
            "Weeks",
            "Months",
            "Years",
          ]
        }
        className="period"
        label=""
        labelNote=""
        name="period-1"
        onChange={[Function]}
        type="dropdown-select"
      />
      <p
        className="expiry-text-mobile"
      >
        To Expiry
      </p>
    </div>
    <p
      className="expiry-text-desktop"
    >
      To Expiry
    </p>
    <button
      className="delete-icon"
      onClick={[Function]}
      type="button"
    >
      <img
        alt="clicked"
        className="addsvg"
        src="delete.svg"
      />
    </button>
  </div>
  <div
    className="reminders-fieldset"
    key="2"
  >
    <Input
      autoComplete="off"
      choices={
        Array [
          Object {
            "label": "Email Reminder",
            "value": "1",
          },
        ]
      }
      className=""
      currentPage={3}
      fetchAllEmailTemplates={[Function]}
      label="Reminder 3"
      labelNote=""
      loading={false}
      name="reminderTemplate-2"
      onChange={[Function]}
      pageCount={4}
      templatesCount={1}
      type="dropdown-select"
    />
    <div
      className="dropdown-input-group"
    >
      <Input
        autoComplete="off"
        className=""
        id={2}
        label="Date"
        labelNote=""
        min={1}
        name="date-2"
        onChange={[Function]}
        type="number"
      />
      <Input
        autoComplete="off"
        choices={
          Array [
            "Days",
            "Weeks",
            "Months",
            "Years",
          ]
        }
        className="period"
        label=""
        labelNote=""
        name="period-2"
        onChange={[Function]}
        type="dropdown-select"
      />
      <p
        className="expiry-text-mobile"
      >
        To Expiry
      </p>
    </div>
    <p
      className="expiry-text-desktop"
    >
      To Expiry
    </p>
    <button
      className="delete-icon"
      onClick={[Function]}
      type="button"
    >
      <img
        alt="clicked"
        className="addsvg"
        src="delete.svg"
      />
    </button>
  </div>
  <AddInputRowButton
    addRowHandler={[Function]}
    buttonClassName="add-reminder__button"
    text="Add a Reminder"
    wrapperClassName="add-reminder__border-dashed"
  />
</fieldset>
`;
